### PR TITLE
Added solution for Project Euler problem 123

### PR DESCRIPTION
**Pull request by ksr1122**
Name: Prime square remainders

Let pn be the nth prime: 2, 3, 5, 7, 11, ..., and
let r be the remainder when (pn−1)^n + (pn+1)^n is divided by pn^2.

For example, when n = 3, p3 = 5, and 43 + 63 = 280 ≡ 5 mod 25.
The least value of n for which the remainder first exceeds 10^9 is 7037.

Find the least value of n for which the remainder first exceeds 10^10.

Reference: https://projecteuler.net/problem=123

reference: #2695

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by ksr1122** [comment]
Review request, @dhruvmanila

**Review by dhruvmanila — COMMENTED** [review]
### Announcement:

@ksr1122

*This message is sent to all pull requests which are **OPEN, NOT MERGED, not containing 'wontfix' label and contains 'Project Euler' or 'Project' or 'Euler' (case doesn't matter) in their title**. If this message doesn't apply to your pull request, please ignore this.*

### Message:

This is to notify all the users submitting a pull request for the Project Euler solution that the [README.md](https://github.com/TheAlgorithms/Python/blob/master/project_euler/README.md) file has been updated with all the style requirements necessary for the directory and solution files. ***Please read through it and make all the necessary changes required to your submission.***

**Review by 53jk1 — APPROVED** [review]

File: project_euler/problem_123/sol1.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+n=4: 2
+n=5: 10p
+
+So it could be simplified as,
+    r = 2pn when n is odd
+    r = 2   when n is even.
+"""
+
+
+def sieve() -> int:
```
Comment (dhruvmanila) [inline project_euler/problem_123/sol1.py:42]: This doesn't return `int`, it returns a generator object. Please make the changes.

File: project_euler/problem_123/sol1.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    Returns a prime number generator using sieve method.
```
Comment (dhruvmanila) [inline project_euler/problem_123/sol1.py:44]: `doctest`

File: project_euler/problem_123/sol1.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    D = {}
+    q = 2
```
Comment (dhruvmanila) [inline project_euler/problem_123/sol1.py:47]: Descriptive variable names

File: project_euler/problem_123/sol1.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    Returns the square remainder with a simplified expression.
+    >>> square_remainder(7, 5)
+    70
+    >>> square_remainder(11, 10)
+    220
+    >>> square_remainder(13, 5)
+    130
+    """
+    return 2 * p * n
```
Comment (dhruvmanila) [inline project_euler/problem_123/sol1.py:71]: You don't need another function for such a small amount of work.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: project_euler/problem_123/sol1.py (hunk 1 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py]: Removed int. Unfortunately I didn't find a way to mention return type for generator.

File: project_euler/problem_123/sol1.py (hunk 2 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py:44]: Added a test to check the type for the return object, which is a generator. Also failed to find a way to test the generated values.

File: project_euler/problem_123/sol1.py (hunk 4 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py]: removed.

File: project_euler/problem_123/sol1.py (hunk 3 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py]: updated.

File: project_euler/problem_123/sol1.py (hunk 1 above)
Comment (dhruvmanila) [reply project_euler/problem_123/sol1.py]: https://docs.python.org/3/library/typing.html#typing.Generator

File: project_euler/problem_123/sol1.py (hunk 2 above)
Comment (dhruvmanila) [reply project_euler/problem_123/sol1.py:44]: Search on how to generate one value at a time using `next` keyword

File: project_euler/problem_123/sol1.py (hunk 1 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py]: I tried it, but it was failing for some reason. Not able to resolve `Generator`. I will give it a another try.

File: project_euler/problem_123/sol1.py (hunk 2 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py:44]: Ya I am already doing it in the function where I call this.

**Comment by ksr1122** [comment]
Tests are failing for some unrelated reason.

File: project_euler/problem_123/sol1.py (hunk 1 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py]: Done.

File: project_euler/problem_123/sol1.py (hunk 2 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py:46]: Done.

File: project_euler/problem_123/sol1.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            while x in factor_map:
+                x += factor
+            factor_map[x] = factor
+        else:
+            factor_map[prime * prime] = prime
+            yield prime
+        prime += 1
+
+
+def solution(limit: int = 1e10) -> int:
```
Comment (dhruvmanila) [inline project_euler/problem_123/sol1.py:78]: This is an incorrect type hint for `limit`:
```console
>>> a = 1e10
>>> type(a)
<class 'float'>
```

File: project_euler/problem_123/sol1.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    >>> next(primes)
+    5
+    >>> next(primes)
+    7
+    >>> next(primes)
+    11
+    >>> next(primes)
+    13
+    """
+    factor_map = {}
```
Comment (dhruvmanila) [inline project_euler/problem_123/sol1.py:63]: Need type annotation for `factor_map`

File: project_euler/problem_123/sol1.py (hunk 2 above)
Comment (dhruvmanila) [reply project_euler/problem_123/sol1.py:46]: This seems to be an incorrect type hint according to : `Generator[YieldType, SendType, ReturnType]`

Your function only `yield` and does not `return`

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: project_euler/problem_123/sol1.py (hunk 2 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py:46]: updated.

File: project_euler/problem_123/sol1.py (hunk 5 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py:78]: changed.

File: project_euler/problem_123/sol1.py (hunk 6 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py:63]: added.

File: project_euler/problem_123/sol1.py (hunk 1 above)
Comment (ksr1122) [reply project_euler/problem_123/sol1.py]: fixed the type.

**Review by dhruvmanila — APPROVED** [review]

**Comment by dhruvmanila** [comment]
Thank you for your contribution!

**Pull request merged**